openmp parallel for, incorrect code generation if loop variable declared private

If a loop iterator variable in a OMP `for` pragma is explicitly declared private, Pythran appears to generate incorrect C++ code.

Test case:

```python

import numpy as np

#pythran export omp_pfor()
def omp_pfor():
    a = np.array([1, 2, 3, 4, 5])

    #omp parallel for private(i)
    for i in range(5):
        a[i] *= i

    return a
```

Without the `private(i)` Pythran correctly generates C++ along the lines of:
```c
    {
      long  __target01234 = 5L;
      #pragma omp parallel for
      for ( i=0L; i < __target01234; i += 1L)
      {
        a[i] *= i;
      }
    }
    return a;
```

However the equivalent with `private(i)` declared, causes Pythran to generate:

```c
    typedef typename pythonic::assignable<__type8>::type __type15;
    __type15 i;
    {
      long  __target01234 = 5L;
      #pragma omp parallel for private(i)
      for ( i=0L; i < __target01234; i += 1L)
      {
        a[i] *= i;
      }
      if (i == __target01234)
      i -= 1L;
    }
    return a;
```

This is obviously problematic as `i` outside the loop is not initialized and for nested loops this extra code prevents OMP loop collapsing since the C++ loops are no longer perfectly nested.

I'm not sure of the best way to fix this, but it seems any declared OMP variables (`private` and `shared`) are included in `openmp_deps` used in the `islocal` scope test [here](https://github.com/serge-sans-paille/pythran/blob/7f3c0c3c198da3e199bfb49f3db82b82483f6031/pythran/backend.py#L593). It makes sense for `shared` OMP variables to not be local, but shouldn't `private` OMP variables be considered local here?,... or I could be misunderstanding the context.

Thanks for reporting, and sorry for the delay. I have difficulties understanding how this is "obviously problematic". If you don't forwrad declare the `i` variable, you end up with

```
omp_pfor.cpp:71:40: error: ‘i’ has not been declared
   71 |       #pragma omp parallel for private(i)
```

So it's correct to forward declare it. The trailing test could be removed though, as it indeed prevents collapsing. Is that what uou meant?

--- a/.github/workflows/parallel.yml
+++ b/.github/workflows/parallel.yml
@@ -42,4 +42,5 @@ jobs:
         export CXX=`echo ${{ matrix.cpp-version }} | sed -e 's/clang/clang++/'`
         export CXXFLAGS="-fopenmp"
         export LDFLAGS="-fopenmp"
+        printf '[compiler]\nldflags=-fopenmp\ncflags=-fopenmp\n' > ~/.pythranrc
         pytest pythran/tests/test_cases.py -v --numprocesses=1

--- a/pythran/backend.py
+++ b/pythran/backend.py
@@ -600,9 +600,14 @@ class CxxFunction(ast.NodeVisitor):
         else:
             # For yield function, upper_bound is globals.
             iter_type = ""
-            # Back one step to keep Python behavior (except for break)
-            loop = [If("{} == {}".format(local_iter, upper_bound),
-                    Statement("{} -= {}".format(local_iter, step)))]
+
+            # Back one step to keep Python behavior (unless the loop index does
+            # not escape)
+            if node.target.id in self.scope[node]:
+                loop = []
+            else:
+                loop = [If("{} == {}".format(local_iter, upper_bound),
+                        Statement("{} -= {}".format(local_iter, step)))]
 
         comparison = self.handle_real_loop_comparison(args, local_iter,
                                                       upper_bound)

--- a/pythran/tests/openmp.legacy/pythran_collapse.py
+++ b/pythran/tests/openmp.legacy/pythran_collapse.py
@@ -1,5 +1,5 @@
 import numpy as np
-def collapse_bug(x, m, btx):
+def collapse_bug0(x, m, btx):
     n = int(np.log2(m))
     lv = np.zeros((x.shape[0], n))
     N = x.shape[0]
@@ -9,8 +9,19 @@ def collapse_bug(x, m, btx):
             lv[i,k] = abs(x[i]- btx[k])
     return lv
 
+def collapse_bug1(x, m, btx):
+    n = int(np.log2(m))
+    lv = np.zeros((x.shape[0], n))
+    N = x.shape[0]
+    #omp parallel for collapse(2) private(i,k)
+    for i in range(N):
+        for k in range(n):
+            lv[i,k] = abs(x[i]- btx[k])
+    return lv
+
+
 def pythran_collapse():
     from random import randint
     n = randint(50, 50)
     x = np.ones(n)
-    return collapse_bug(x, 5, 3. * x)
+    return collapse_bug0(x, 5, 3. * x), collapse_bug1(x, 5, 3. * x)

